common/searchattribute/defs.go

```diff
@@ -29,75 +29,72 @@ import (
-		TaskQueue:         {},
-		HistoryLength:     {},
-		ExecutionDuration: {},
-
-		MemoEncoding:      {},
-		Memo:              {},
-		VisibilityTaskKey: {},
+	// system are internal search attributes which are passed and stored as separate fields.
+	system = map[string]enumspb.IndexedValueType{
+		WorkflowID:           enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+		RunID:                enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+		WorkflowType:         enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+		StartTime:            enumspb.INDEXED_VALUE_TYPE_DATETIME,
+		ExecutionTime:        enumspb.INDEXED_VALUE_TYPE_DATETIME,
+		CloseTime:            enumspb.INDEXED_VALUE_TYPE_DATETIME,
+		ExecutionStatus:      enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+		TaskQueue:            enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+		HistoryLength:        enumspb.INDEXED_VALUE_TYPE_INT,
+		ExecutionDuration:    enumspb.INDEXED_VALUE_TYPE_INT,
+		StateTransitionCount: enumspb.INDEXED_VALUE_TYPE_INT,
 	}
 
-	// system are default internal system search attributes which are shown to the users.
-	// Note: NamespaceID is not included because it is not exposed.
-	system = map[string]enumspb.IndexedValueType{
-		WorkflowID:            enumspb.INDEXED_VALUE_TYPE_KEYWORD,
-		RunID:                 enumspb.INDEXED_VALUE_TYPE_KEYWORD,
-		WorkflowType:          enumspb.INDEXED_VALUE_TYPE_KEYWORD,
-		StartTime:             enumspb.INDEXED_VALUE_TYPE_DATETIME,
-		ExecutionTime:         enumspb.INDEXED_VALUE_TYPE_DATETIME,
-		CloseTime:             enumspb.INDEXED_VALUE_TYPE_DATETIME,
-		ExecutionStatus:       enumspb.INDEXED_VALUE_TYPE_KEYWORD,
-		TaskQueue:             enumspb.INDEXED_VALUE_TYPE_KEYWORD,
-		HistoryLength:         enumspb.INDEXED_VALUE_TYPE_INT,
-		ExecutionDuration:     enumspb.INDEXED_VALUE_TYPE_INT,
-		StateTransitionCount:  enumspb.INDEXED_VALUE_TYPE_INT,
+	// predefined are internal search attributes which are passed and stored in SearchAttributes object together with custom search attributes.
+	predefined = map[string]enumspb.IndexedValueType{
 		TemporalChangeVersion: enumspb.INDEXED_VALUE_TYPE_KEYWORD,
 		BinaryChecksums:       enumspb.INDEXED_VALUE_TYPE_KEYWORD,
 		BatcherNamespace:      enumspb.INDEXED_VALUE_TYPE_KEYWORD,
-		BarcherUser:           enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+		BatcherUser:           enumspb.INDEXED_VALUE_TYPE_KEYWORD,
 	}
 
-	namespaceIDType = enumspb.INDEXED_VALUE_TYPE_KEYWORD
+	// reserved are internal field names that can't be used as search attribute names.
+	reserved = map[string]struct{}{
+		NamespaceID:       {},
+		MemoEncoding:      {},
+		Memo:              {},
+		VisibilityTaskKey: {},
+	}
 )
 
-// IsReservedField return true if field name is system reserved.
-func IsReservedField(fieldName string) bool {
-	_, ok := reservedFields[fieldName]
-	return ok
+// IsReserved return true if field name is system reserved and can't be used as custom search attribute name.
+func IsReserved(fieldName string) bool {
+	if _, ok := system[fieldName]; ok {
```

Comment:
This is a good question. I don't have strong answer for it. Probably all new system search attributes should start with `Temporal` and we should advertise this upfront. Like "Never prefix your custom search attributes with `Temporal` because it may collides with system names in the future".